File: Bwl.Imaging/Tools/ResourceCollector.vb
Imports System.Collections.Concurrent
Imports System.Threading

Public Interface ICollectedResource
    Inherits IComparable(Of ICollectedResource)
    ReadOnly Property CollectTimeUtc As Date
    Function TryCollect() As Boolean
End Interface

Public NotInheritable Class ResourceCollector(Of T)
    Private Shared ReadOnly _resources As LinkedList(Of ICollectedResource) = New LinkedList(Of ICollectedResource)()
    Private Shared ReadOnly _incoming As ConcurrentQueue(Of ICollectedResource) = New ConcurrentQueue(Of ICollectedResource)()
    Private Shared _count, _procTimeMs As Long
    Public Shared ReadOnly Property Count As Long
        Get
            Return Interlocked.Read(_count)
        End Get
    End Property
    Public Shared ReadOnly Property ProcTimeMs As Long
        Get
            Return Interlocked.Read(_procTimeMs)
        End Get
    End Property
    Public Shared PeriodMs As Double = 250.0
    Shared Sub New()

        Call Task.Run(Function() ResourceCollectorTask().ConfigureAwait(False))
    End Sub
    Public Shared Sub Add(item As ICollectedResource)
        _incoming.Enqueue(item)
    End Sub
    Private Shared Async Function ResourceCollectorTask() As Task
        Dim sw = New Stopwatch()
        Dim node As LinkedListNode(Of ICollectedResource) = Nothing
        Dim item As ICollectedResource = Nothing
        While True
            sw.Restart()
            Dim nowUtc = Date.UtcNow
            While _incoming.Any()
                _incoming.TryDequeue(item)
                node = _resources.Last
                While node IsNot Nothing AndAlso node.Value.CompareTo(item) > 0
                    node = node.Previous
                End While
                If node IsNot Nothing Then
                    _resources.AddAfter(node, item)
                Else
                    _resources.AddFirst(item)
                End If
                Interlocked.Increment(_count)
            End While
            node = _resources.First
            While node IsNot Nothing
                Dim current = node
                node = node.Next
                If current.Value.CollectTimeUtc > nowUtc Then Exit While
                Dim collected As Boolean

                Try
                    collected = current.Value.TryCollect()
                Catch
                    collected = False
                End Try
                If collected Then
                    _resources.Remove(current)
                    Interlocked.Decrement(_count)
                End If
            End While
            sw.Stop()
            _procTimeMs = sw.ElapsedMilliseconds
            Await Task.Delay(TimeSpan.FromMilliseconds(Math.Max(PeriodMs - sw.Elapsed.TotalMilliseconds, 0))).ConfigureAwait(False)
        End While
    End Function
End Class

File: Bwl.Imaging.Skia/Tools/ResourceCollector.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace Bwl.Imaging.Skia
{

    public interface ICollectedResource : IComparable<ICollectedResource>
    {
        DateTime CollectTimeUtc { get; }
        bool TryCollect();
    }

    public sealed class ResourceCollector<T>
    {
        private static readonly LinkedList<ICollectedResource> _resources = new LinkedList<ICollectedResource>();
        private static readonly ConcurrentQueue<ICollectedResource> _incoming = new ConcurrentQueue<ICollectedResource>();
        private static long _count, _procTimeMs;
        public static long Count
        {
            get
            {
                return Interlocked.Read(ref _count);
            }
        }
        public static long ProcTimeMs
        {
            get
            {
                return Interlocked.Read(ref _procTimeMs);
            }
        }
        public static double PeriodMs = 250.0d;
        static ResourceCollector()
        {

            Task.Run(() => ResourceCollectorTask().ConfigureAwait(false));
        }
        public static void Add(ICollectedResource item)
        {
            _incoming.Enqueue(item);
        }
        private static async Task ResourceCollectorTask()
        {
            var sw = new Stopwatch();
            LinkedListNode<ICollectedResource> node = null;
            ICollectedResource item = null;
            while (true)
            {
                sw.Restart();
                var nowUtc = DateTime.UtcNow;
                while (_incoming.Any())
                {
                    _incoming.TryDequeue(out item);
                    node = _resources.Last;
                    while (node is not null && node.Value.CompareTo(item) > 0)
                        node = node.Previous;
                    if (node is not null)
                    {
                        _resources.AddAfter(node, item);
                    }
                    else
                    {
                        _resources.AddFirst(item);
                    }
                    Interlocked.Increment(ref _count);
                }
                node = _resources.First;
                while (node is not null)
                {
                    var current = node;
                    node = node.Next;
                    if (current.Value.CollectTimeUtc > nowUtc)
                        break;
                    bool collected;

                    try
                    {
                        collected = current.Value.TryCollect();
                    }
                    catch
                    {
                        collected = false;
                    }
                    if (collected)
                    {
                        _resources.Remove(current);
                        Interlocked.Decrement(ref _count);
                    }
                }
                sw.Stop();
                _procTimeMs = sw.ElapsedMilliseconds;
                await Task.Delay(TimeSpan.FromMilliseconds(Math.Max(PeriodMs - sw.Elapsed.TotalMilliseconds, 0d))).ConfigureAwait(false);
            }
        }
    }
}